app/src/main/java/bg/zahov/app/ui/exercise/ExerciseViewModel.kt

```diff
@@ -131,55 +131,61 @@ class ExerciseViewModel(
+     * Checks if an exercise matches the search string.
      *
-     * @param filter The list of filters to apply.
-     * @param exercises The list of exercises to filter.
-     * @param searchString The search query to match against exercise names.
-     * @return A filtered list of exercises that match the filters and search query.
+     * @param exercise The exercise to test.
+     * @param searchString The search query to match against the exercise name. Defaults to the current search string in the state.
+     * @return `true` if the search string is blank or the exercise name contains the search string (case-insensitive).
      */
-    private fun getFiltered(
-        filter: List<FilterItem> = _exerciseData.value.filters,
-        exercises: List<ExerciseData> = _exerciseData.value.exercises,
+    private fun matchesSearch(
+        exercise: ExerciseData,
         searchString: String = _exerciseData.value.search
-    ): List<ExerciseData> {
-        return exercises.map { exercise ->
-            if ((filter.isEmpty() ||
-                        filter.any { filterWrapper ->
-                            matchesFilter(exercise, filterWrapper.filter)
-                        })
-                && (searchString.isBlank() ||
-                        exercise.name.contains(searchString, ignoreCase = true))
-            ) {
-                exercise.copy(toShow = true)
-            } else {
-                exercise.copy(toShow = false)
-            }
-        }
+    ): Boolean {
+        return searchString.isBlank() || exercise.name.contains(searchString, ignoreCase = true)
     }
 
     /**
-     * Checks if the given exercise matches the specified filter.
+     * Checks if an exercise matches the mandatory filters (body part and category).
      *
-     * @param exercise The exercise to check against the filter.
-     * @param filter The filter to check.
-     * @return True if the exercise matches the filter; otherwise, false.
+     * @param exercise The exercise to test.
+     * @param filterList The list of filters to evaluate. Defaults to the current filters in the state.
+     * @return `true` if the exercise matches both the body part and category filters, or if those filters are not present.
      */
-    private fun matchesFilter(
+    private fun matchesAnyFilter(
         exercise: ExerciseData,
-        filter: Filter
+        filterList: List<FilterItem> = _exerciseData.value.filters
     ): Boolean {
-        return when (filter) {
-            is Filter.CategoryFilter -> {
-                exercise.category == filter.category
-            }
+        val categoryFilters = filterList.filter { it.filter is Filter.CategoryFilter }
+        val bodyPartFilters = filterList.filter { it.filter is Filter.BodyPartFilter }
 
-            is Filter.BodyPartFilter -> {
-                exercise.bodyPart == filter.bodyPart
-            }
+        val categoryMatch = categoryFilters.isEmpty() || categoryFilters.any { filterItem ->
```

Comment:
If the list is empty then the second check `categoryFilters.any` already covers that case. `.any()` uses `isEmpty()` inside of it if you check the source code. Therefore the `isEmpty()` check is redundant.

Hypothetically If you want to add a condition that can skip the `categoryFilters.any` check then you can instead use an AND operator, e.g.

```suggestion
        val categoryMatch = categoryFilters.isEmpty().not() && categoryFilters.any { filterItem ->
```

That way if the first check is false the `categoryFilters.any` won't be executed and computational power can be conserved. As mentioned this is not needed in this case, but it's useful in most cases where heavy operations are needed.


The same comment applies for the  `val bodyPartMatch`.

Reply:
I will update the kdoc to not be so vague

Reply:
Yes, my bad on that.